collector/src/main/java/com/bitmovin/analytics/retryBackend/RetryBackend.kt

```diff
@@ -0,0 +1,105 @@
+package com.bitmovin.analytics.retryBackend
+
+import android.os.Handler
+import android.os.SystemClock
+import android.util.Log
+import com.bitmovin.analytics.data.AdEventData
+import com.bitmovin.analytics.data.Backend
+import com.bitmovin.analytics.data.CallbackBackend
+import com.bitmovin.analytics.data.EventData
+import java.io.IOException
+import java.lang.Exception
+import java.net.ConnectException
+import java.net.SocketTimeoutException
+import java.net.UnknownHostException
+import java.util.Date
+import okhttp3.Call
+import okhttp3.Response
+import okhttp3.internal.http2.StreamResetException
+
+class RetryBackend(private val next: CallbackBackend, val handler: Handler) : Backend {
+
+    private val TAG = "RetryBackend"
+    private var retryDateToken: Date? = null
+    private val retryQueue = RetryQueue()
+    fun getNextScheduledTime() = retryQueue.getNextScheduleTime()
+
+    override fun send(eventData: EventData) {
+        scheduleSample(RetrySample(eventData, 0, Date(), 0))
+    }
+
+    override fun sendAd(eventData: AdEventData) {
+        scheduleSample(RetrySample(eventData, 0, Date(), 0))
+    }
+
+    private fun scheduleSample(retrySample: RetrySample<Any>) {
+
+        if (retrySample.eventData is EventData) {
+            Log.d(TAG, "sending sample ${retrySample.eventData?.sequenceNumber} retry ${retrySample.retry}")
+            retrySample.eventData.retryCount = retrySample.retry
+            this.next.send(retrySample.eventData, object : RetryCallback {
+                override fun onFailure(call: Call, e: IOException) {
+                    if (e is SocketTimeoutException || e is ConnectException || e is StreamResetException || e is UnknownHostException) {
+                        call.cancel()
+                        retryQueue.addSample(retrySample)
+                        processQueuedSamples()
+                    }
+                }
+
+                override fun onResponse(call: Call, response: Response) {
+                }
+            })
+        } else if (retrySample.eventData is AdEventData) {
+            retrySample.eventData.retryCount = retrySample.retry
+            this.next.sendAd(retrySample.eventData, object : RetryCallback {
+                override fun onFailure(call: Call, e: IOException) {
+                    if (e is SocketTimeoutException || e is ConnectException || e is StreamResetException || e is UnknownHostException) {
+                        call.cancel()
+                        retryQueue.addSample(retrySample)
+                        processQueuedSamples()
+                    }
+                }
+
+                override fun onResponse(call: Call, response: Response) {
+                }
+            })
+        }
+    }
+
+    @Synchronized
+    fun processQueuedSamples() {
+        try {
+            val nextScheduledTime = getNextScheduledTime()
+            if (nextScheduledTime != null) {
```

Comment:
this should do the same and avoid nesting:
```suggestion
            if (nextScheduledTime == null || retryDateToken == nextScheduledTime) { 
            return
            }
```